src/java.base/share/classes/java/lang/Runtime.java

```diff
@@ -83,10 +83,12 @@ private Runtime() {}
      * until they finish.  Once this is done the virtual machine
      * {@linkplain #halt halts}.
      *
-     * <p> If this method is invoked from more than one thread the status code
-     * of the first invocation will be used; the status codes from other invocations
-     * will be ignored. If this method is invoked from a shutdown hook the system
-     * will deadlock.
+     * <p> Shutdown is serialized such that only one invocation will run
+     * shutdown hooks and terminate the VM with the given status code. That
+     * invocation may be initiated via platform specific signal handlers. All
```

Comment:
Why are we mentioning signal handlers here? How is that relevant?

Reply:
Signal handlers for eg SIGTERM invoke Shutdown.shutdown. That method holds the same lock as Shutdown.exit and runs shutdown hooks. So if a signal handler triggers shutdown, and before the system halts Runtime.exit is invoked, the status passed to Runtime.exit will be ignored.

Reply:
First, the signal handlers actually invoke Shutdown.exit, not Shutdown.shutdown  - see Terminator.setup(). (If they did the latter, like DestroyJavaVM, then an exit(status) call from another thread could still do the final VM halt(status)

But now you are opening a can of worms. There are multiple ways for the VM to initiate termination - are you going to try and describe here how all of them potentially interact?  

What you are really stating here is that other parts of the JDK can invoke System.exit, but that is for them to specify where such things are specified, it isn't for exit() to try and list them all. All we have to do here is describe how multiple calls to exit() behave.

Reply:
Aha! I think I misread the comment on Shutdown.shutdown and had it stuck in my head that signals would exit through that method. Thanks for the explanation, it makes a lot more sense now.

> All we have to do here is describe how multiple calls to exit() behave.

The thing that is still missing to me is that an application developer may have a single call to System.exit, yet the status code they pass to it may not be the one the jvm exits with.

Reply:
Yes and the problem is? If the user kills the VM by other means; or the termination of the last non-daemon thread races with a call to System.exit then the developer doesn't really have to right to expect anything in relation to that one call.